Sources/SourceKitLSP/WorkDoneProgressManager.swift

```diff
@@ -0,0 +1,57 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LanguageServerProtocol
+import SKSupport
+
+/// Represents a single `WorkDoneProgress` task that gets communicated with the client.
+///
+/// The work done progress is started when the object is created and ended when the object is destroyed.
+/// In between, updates can be sent to the client.
+final class WorkDoneProgressManager {
+  private let token: ProgressToken
+  private let queue = AsyncQueue<Serial>()
+  private let server: SourceKitServer
+
+  init(server: SourceKitServer, title: String, message: String? = nil) {
+    self.token = .string("WorkDoneProgress-\(UUID())")
+    self.server = server
+    queue.async { [server, token] in
+      await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
+        _ = server.client.send(CreateWorkDoneProgressRequest(token: token)) { result in
+          continuation.resume()
+        }
+      }
+      await server.sendNotificationToClient(
+        WorkDoneProgress(token: token, value: .begin(WorkDoneProgressBegin(title: title, message: message)))
+      )
+    }
+  }
+
+  func update(message: String? = nil, percentage: Int? = nil) {
```

Comment:
Is it just a spinny/something is percentage is nil?

Reply:
Yes, it looks similar to this building entry.

<img width="85" alt="Screenshot 2024-02-28 at 09 24 28" src="https://github.com/apple/sourcekit-lsp/assets/4062178/ce571340-c185-4e0e-9e00-203f14d1ea40">